src/main/java/org/jboss/aerogear/keycloak/metrics/PrometheusExporter.java

```diff
@@ -1,31 +1,37 @@
 package org.jboss.aerogear.keycloak.metrics;
 
+import com.beust.jcommander.internal.Lists;
 import io.prometheus.client.CollectorRegistry;
 import io.prometheus.client.Counter;
 import io.prometheus.client.exporter.common.TextFormat;
 import io.prometheus.client.hotspot.DefaultExports;
+import org.jboss.logging.Logger;
 import org.keycloak.events.Event;
 import org.keycloak.events.EventType;
 import org.keycloak.events.admin.AdminEvent;
 import org.keycloak.events.admin.OperationType;
+import org.keycloak.events.admin.ResourceType;
 
 import java.io.*;
 import java.util.HashMap;
+import java.util.List;
 import java.util.Map;
 
 public final class PrometheusExporter {
 
+    private final static Logger logger = Logger.getLogger(PrometheusExporter.class);
+
     private final static String USER_EVENT_PREFIX = "keycloak_user_event_";
     private final static String ADMIN_EVENT_PREFIX = "keycloak_admin_event_";
     private final static String PROVIDER_KEYCLOAK_OPENID = "keycloak";
 
     private final static PrometheusExporter INSTANCE = new PrometheusExporter();
 
-    // these fields are package private by on purpose
-    final Map<String, Counter> counters = new HashMap<>();
-    final Counter totalLogins;
-    final Counter totalFailedLoginAttempts;
-    final Counter totalRegistrations;
+    private final UserEventChain userEventChain = new UserEventChain();
```

Comment:
@aliok as far as i understand it the purpose of the UserEventChain is to preallocate a number of handlers for well known events and then reuse those and create counters on the fly for all other events?

Reply:
@pb82 
I wanted to unify where to store the counters. In order to do that, I needed to put the counters for well known events to the same map with other counters.

However, when I was doing that, I didn't want to have a specific `recordLogin` etc. methods so that I can move some business logic from MetricsListener to PrometheusExporter.
I ended up creating a chain that handles well known events. Events that don't fall into that category are not handled by that chain and considered generic.

That chain there creates counters for the well known events and also processes those events.

If we want to have a new event as a "well known" event in the future, we just need to create a handler and add it to the chain. No need for adding a new case into a switch statement nor a new `recordFoo` method.

Reply:
Ok, understand 👍